File:  Kotlin Practice/src/programs/Compress.java
package programs;

public class Compress {
    public static void main(String[] args) {
        System.out.println(compress(new char[] {'a', 'a', 'b', 'b', 'c', 'c', 'c'}));
        System.out.println(compress(new char[] {'a'}));
        System.out.println(compress(new char[] {'a', 'b', 'b', 'b', 'b', 'b', 'b', 'b', 'b', 'b', 'b', 'b', 'b'}));
    }

    public static int compress(char[] chars) {

        if (chars.length == 1) {
            return 1;
        }

        StringBuilder s = new StringBuilder();
        char character = chars[0];
        int charCount = 1;

        for (int i = 1; i <= chars.length - 1; i++) {
            if (chars[i] == character) {
                charCount++;

                if (i == chars.length - 1) {
                    s.append(character);
                    if (charCount > 1) {
                        s.append(Integer.toString(charCount));
                    }
                }
            } else {
                s.append(character);
                if (charCount > 1) {
                    s.append(Integer.toString(charCount));
                }

                character = chars[i];
                charCount = 1;
            }
        }

        chars = s.toString().toCharArray();
        chars = new char[20];
        System.out.println(chars);
        return chars.length;
    }

}

File:  Kotlin Practice/src/programs/compress.kt
package programs

fun main() {
  println(compress(charArrayOf('a','a','b','b','c','c','c')))
  println(compress(charArrayOf('a')))
  println(compress(charArrayOf('a','b','b','b','b','b','b','b','b','b','b','b','b')))
}

fun compress(chars: CharArray): Int {

    if(chars.size == 1){
        return 1
    }

    val s = StringBuilder()
    var char = chars[0]
    var charCount = 1

    for(i in 1..chars.lastIndex){
        if(chars[i] == char){
            charCount++

            if(i == chars.lastIndex){
                s.append(char)
                if(charCount > 1){
                    s.append(charCount.toString())
                }
            }
        }else{
            s.append(char)
            if(charCount > 1){
                s.append(charCount.toString())
            }

            if(i == chars.lastIndex && chars[i] != char){
                s.append(chars[i])
            }

            char = chars[i]
            charCount = 1
        }
    }

    for(i in chars.indices){
       if(i > s.lastIndex){
           chars[i] = ' '
       }else{
           chars[i] = s[i]
       }
    }

    println("Chars : ${chars.contentToString()}")
    return s.length
}